lib: os: add CRC shell command for integrity verification

cfriedt opened the pull request:
Several Zephyr shell commands are used for transferring data over a possibly unreliable connection such as a UART into either memory or flash. For example, `flash load` and `devmem load`.

Make the CRC functions available so that they can be used to verify the integrity of data transferred over possibly unreliable connections.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
lib: os: add CRC shell command for integrity verification

There are several Zephyr shell commands that are used for
transferring data over a possibly unreliable connection such as a
UART into either memory or flash. For example, `flash load` and
`devmem load`.

Make the CRC functions available so that they can be used to
verify the integrity of data transferred over possibly
unreliable connections.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

carlescufi reviewed (dismissed):
Very nice, thanks!

cfriedt commented on lib/os/crc_shell.c:
```diff
@@ -0,0 +1,185 @@
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(crc_types); ++i) {
+		printk("compare: %s %s\n", s, crc_types[i]);
```
Shoot - i just realized I left this here m(

cfriedt replied:
@carlescufi - can you please re-ack?

cfriedt commented:
* removed stray `printk()` call

jukkar commented on lib/os/crc_shell.c:
```diff
@@ -0,0 +1,184 @@
+};
+
+/* These must match the order of enum crc_type */
+static const char *const crc_types[] = {
+	"16", "8", "16_ccitt", "16_itu_t", "16_ansi", "32_ieee", "32_c", "8_ccitt", "7_be",
+};
```
This could be slightly error prone if one adds new crc enum values but misses the corresponding string.
We could do it a bit more safely with this
```
static const char *const crc_types[] = {
	[CRC8] = "8",
	[CRC16] = "16",
...
};
```

cfriedt replied:
Done!

Commit:
include: sys: crc: add crc_by_type inline dispatch function

Rather than hard-coding calls to each individual CRC function, it
may be more convenient to specify the crc algorithm to use
parametrically.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
* indexed string table with enums
* moved dispatch code to `crc_by_type()` in `<zephyr/sys/crc.h>`

jukkar reviewed (approved):
LGTM
